### DISCUSSION TITLE
Keyboard on demand

### DISCUSSION THREAD

**Discussion by tomelgato** [davetcc/tcMenu#314] [General]
Hello Dave, one more thing ;)

I use the 5x4 keypad, 20x4 LCD together with the AccelStepper library. When moving the stepper motor i dont call taskManager.runLoop(), otherwise the performance on the Due is not sufficient. So far so good.

But now i need a function where i need to poll the keyboard (nothing menu related) and drive the stepper at the same time. The plan is to poll the keyboard directly. So the Q is if i just can steal the keyboard pins or if i have to close/open them via iOAbstraction/taskmanager/keyboard.

**Comment by davetcc** [comment]
I'm always surprised that nobody seems to use the inbuilt ARM SAM hardware PWM or interrupt support. It seems to me that either of these would control stepper motors much more effectively. But that's an entirely different discussion outside of tcMenu.

The short answer is yes you could directly read it, but there may be another way. Reading a matrix keyboard is slow, and involves many steps, the [guide on keyboards explains it](https://www.thecoderscorner.com/products/arduino-libraries/io-abstraction/matrix-keyboard-keypad-manager/). However, that aside, you could call `exec()` on the keyboard manager, and it would service the keyboard in the usual way without doing anything else. You could even try reducing the stabilization delay in the exec method until you get failures. In your case though, if you did that you'd need to replace the yieldForMicros(..) with delayMicroseconds(..). If this worked, I'd make that change behind a permanent build flag.

**Reply by tomelgato** [reply]
I think the stepper libraries started as uno project ;)

The exec() is an excellent solution. Will try that, should be fast enough. If not i still have the plan B doing a own routine which checks just the 3 buttons i need (they are in one row so it would be quite quick).

Just dont understand your last sentence about the permanent build flag, what do you mean?

**Reply by davetcc** [reply]
Make sure you replace the delayForMicros call with delayMicroseconds in the keyboard manager exec call as otherwise you’re calling runLoop indirectly. 

What I mean is that If that works then I’ll make it configurable in the next version.

**Reply by tomelgato** [reply]
Ok i try and give you feedback. Thanks a lot!

**Reply by tomelgato** [reply]
I think a second exec call like exec_external() which would not call the loop and use Micros would be better then do do some flags, no?

**Reply by davetcc** [reply]
What would be best is to refactor it slightly and have a method something like readGrid(int whichLine) that allowed more control for these situations.

**Reply by tomelgato** [reply]
Yes that would be perfect. Line / Row ... for people using a keypad to control something in X/Y direction. Lol, usual featuritis is coming to effect. ;)

**Reply by tomelgato** [reply]
Ping ;) See below

**Comment by tomelgato** [comment]
Hello Dave! Finally i could do some speed tests. 

taskManager.runLoop() is somewhere in my main file, its not full speed cause some other stuff is running too.

The 2nd measurement is running in a small sub loop without calling taskManager.runLoop(), instead i poll the keyboard with keyboard.exec();  


```
taskManager.yieldForMicros(500);
//delayMicroseconds(500);

taskManager.runLoop(): 184971 Loops/sec
keyboard.exec(): 237798 Loops/sec
---
// taskManager.yieldForMicros(500); 
delayMicroseconds(500);

taskManager.runLoop(): 196871 Loops/sec
keyboard.exec(): 242423 Loops/sec
---
taskManager.yieldForMicros(250); 
//delayMicroseconds(250);

taskManager.runLoop():	192256 Loops/sec
keyboard.exec(): 243672 Loops/sec
—
//taskManager.yieldForMicros(250); 
delayMicroseconds(250);

taskManager.runLoop(): 201026 Loops/sec
keyboard.exec(): 247453 Loops/sec
```

**Reply by tomelgato** [reply]
So altogether its possible to get 10k loops more on the Arduino DUE when not calling the taskmanager indirectly and reducing the delay to 250us
